### PR TITLE
Arbitrary depth rewinding in Optimism forkchoice

### PR DESCRIPTION
**Pull request by jyellick**
The current forkchoice code will not allow re-orgs to canonical blocks. This is consistent with the Ethereum specification, but is incorrect for Optimism.  This change adds the ability to re-org a chain via a forkchoice to a canonical block.

### CODE REVIEW

**Comment by ImTei** [comment]
@jyellick Can you provide more details about the change, such as optimism specs or corresponding op-geth impl?
I think this issue was handled in the past(https://github.com/testinprod-io/op-erigon/pull/8, https://github.com/testinprod-io/op-erigon/pull/16), but it appears to have become an issue again due to a change in upstream code. I wanna get confirm if this PR is the same issue.

File: turbo/engineapi/engine_server.go [hunk 1]
```diff
@@ -468,14 +474,6 @@ func (s *EngineServer) forkchoiceUpdated(ctx context.Context, forkchoiceState *e
...
 	if payloadAttributes != nil {
 		timestamp := uint64(payloadAttributes.Timestamp)
-		// op-node sends V3 before Cancun, without beacon root.
-		// so we must allow V3 before cancun to support forkChoiceUpdated backward compatibility
-		//if !s.config.IsCancun(timestamp) && version >= clparams.DenebVersion { // V3 before cancun
-		//	if payloadAttributes.ParentBeaconBlockRoot == nil {
-		//		return nil, &rpc.InvalidParamsError{Message: "Beacon Root missing"}
-		//	}
-		//	return nil, &rpc.UnsupportedForkError{Message: "Unsupported fork"}
-		//}
```
Comment (ImTei) [inline turbo/engineapi/engine_server.go]: Why did you remove this? this issue was fixed in the [op-node](https://github.com/ethereum-optimism/optimism/releases/tag/v1.5.1) and we should restore this check.

Comment (jyellick) [reply turbo/engineapi/engine_server.go]: Removed as it seemed like cleanup -- I thought there would be a TODO or comment if it was meant to be added back.  I've left it as is.

**Comment by jyellick** [comment]
> I think this issue was handled in the past(https://github.com/testinprod-io/op-erigon/pull/8, https://github.com/testinprod-io/op-erigon/pull/16), but it appears to have become an issue again due to a change in upstream code. I wanna get confirm if this PR is the same issue.

Correct, this is that same issue referenced in those others.  When Erigon refactored the code to better support their internal consensus layer, it broke the re-org path for op-erigon.

Adding this change fixes dozens of broken tests in the op-e2e.

File: turbo/engineapi/engine_server.go [hunk 2]
```diff
@@ -496,6 +502,22 @@ func (s *EngineServer) forkchoiceUpdated(ctx context.Context, forkchoiceState *e
...
+		// {payloadStatus: {status: VALID, latestValidHash: forkchoiceState.headBlockHash, validationError: null}, payloadId: null}.
+		// We skip this check in the Optimism case as Optimism allows arbitrary
+		// depth re-orgs
+
+		s.logger.Warn("Skipping payload building because forkchoiceState.headBlockHash is not the head of the canonical chain",
+			"forkChoice.HeadBlockHash", forkchoiceState.HeadHash, "headHeader.Hash", headHeader.Hash())
+		return &engine_types.ForkChoiceUpdatedResponse{PayloadStatus: status}, nil
+	}
+	log.Debug("Continuing EngineForkChoiceUpdated", "headNumber", headHeader.Number, "headHash", headHeader.Hash(), "numDeposits", len(payloadAttributes.Transactions))
+
```
Comment (ImTei) [inline turbo/engineapi/engine_server.go]: This check does not exists in upstream erigon anymore (https://github.com/ledgerwatch/erigon/pull/8982). So actually we don't need this change.

Comment (jyellick) [reply turbo/engineapi/engine_server.go]: Fixed

**Review by ImTei — APPROVED** [review]
Thanks for founding this issue & fix!
